src/quota_aggregator_impl.cc

```diff
@@ -258,20 +237,47 @@ ::google::protobuf::util::Status QuotaAggregatorImpl::FlushAll() {
 
 // OnCacheEntryDelete will be called behind the cache_mutex_
 // no need to consider locking at this point
+// This is called when an cached item is older than refresh_interval
+// as SetAgeBasedEviction() is called at the constuctor.
 void QuotaAggregatorImpl::OnCacheEntryDelete(CacheElem* elem) {
-  if (in_flush_all_ == false && ShouldDrop(*elem) == false) {
-    // insert the element back to the cache
+  if (in_flush_all_ || ShouldDrop(*elem)) {
+    delete elem;
+    return;
+  }
+
+  if (elem->in_flight()) {
+    // This item is still calling the server, add it back to the cache
+    // to wait for the response.
     cache_->Insert(elem->signature(), elem, 1);
+    return;
+  }
 
-    if (elem->in_flight() == false && elem->is_aggregated()) {
-      elem->set_in_flight(true);
-      elem->set_last_refresh_time(SimpleCycleTimer::Now());
-      AddRemovedItem(elem->ReturnAllocateQuotaRequestAndClear(
-        service_name_, service_config_id_));
+  // For an aggregated item, send the aggregated cost to the server.
+  // For an negative item, send CHECK_ONLY to check available quota.
+  if (elem->is_aggregated() || !elem->is_positive_response()) {
+    elem->set_in_flight(true);
+    elem->set_last_refresh_time(SimpleCycleTimer::Now());
+    auto request = elem->ReturnAllocateQuotaRequestAndClear(
+        service_name_, service_config_id_);
+    if (!elem->is_positive_response()) {
+      request.mutable_allocate_operation()->
+          set_quota_mode(
+              QuotaOperation_QuotaMode::QuotaOperation_QuotaMode_CHECK_ONLY);
     }
-  } else {
-    delete elem;
+    // Insert the element back to the cache
+    // This is important for negative items to reject new requests.
+    cache_->Insert(elem->signature(), elem, 1);
+    // AddRemovedItem function name is misleading, it actually calls
+    // transport function to send the request to server.
+    AddRemovedItem(request);
+    return;
   }
+
+  // The postive and non-aggregated items reach here, they can be
```

Comment:
But this is dropping positive elements before the expiration interval. Will this not cause extra requests to the quota server?

Reply:
Good catch.  If we remove these items,  it will cause more quota calls to server since there will be more cache miss. But our goal is not to reduce number of calls,  our goal is to reduce latency with limited memory usage.  I still prefer to remove them

Reply:
I changed my mind.  we may stick to cache expiration definition,  keep these empty items in the cache.   With reduced expiration time to 1 minutes, and we have 10K cache entries, we should not need to worry about the cache issue.  so we can try to reduce quota call.   

So I will add these items back.  
Then, only change in this pr is to send CHECK_ONLY calls for negative items.